### PR TITLE
Realtime session: Retry sessionStart hit only for recoverable URL and HTTP errors

### PR DESCRIPTION
**Pull request by addb**
Refactored code to make it clean, log all the scenarios. Updated test utils and added tests.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [ ] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

File: AEPMedia/Sources/MediaCollection/MediaRealTimeSession.swift [hunk 1]
```diff
@@ -178,18 +179,59 @@ class MediaRealTimeSession: MediaSession {
...
+
+        guard sessionStartRetryCount < MediaRealTimeSession.MAX_ALLOWED_FAILURE else {
+            Log.trace(label: Self.LOG_TAG, "[\(Self.CLASS_NAME)<\(#function)>] - [Session (\(self.id))] \(eventType) request resulted in error but since it surpassed the retry limit:(\(MediaRealTimeSession.MAX_ALLOWED_FAILURE), the request will be dropped.")
+            return false
+        }
+
+        guard let connection = connection else {
+            // should reach here when no connection (http, url) error
+            // retry for case when error occured while parsing sessionStart response for sessionId
+            Log.trace(label: Self.LOG_TAG, "[\(Self.CLASS_NAME)<\(#function)>] - [Session (\(self.id))] Unable to extract sessionId from sessionStart request response, the request will be retried.")
```
Comment (kevinlind) [inline AEPMedia/Sources/MediaCollection/MediaRealTimeSession.swift:208]: nit: maybe state the connection is nil here in the logs, such as "... from sessionStart request response as the connection is nil, the request will be retried."

File: AEPMedia/Sources/MediaCollection/MediaRealTimeSession.swift [hunk 2]
```diff
@@ -178,18 +179,59 @@ class MediaRealTimeSession: MediaSession {
...
+            if NetworkServiceConstants.RECOVERABLE_ERROR_CODES.contains(responseCode) {
+                Log.debug(label: Self.LOG_TAG, "[\(Self.CLASS_NAME)<\(#function)>] - [Session(\(self.id))] \(eventType) request failed with recoverable HTTP error:(\(connection.responseMessage ?? "")) response code:(\(responseCode)). Request will be retried.")
+                return true // retry for recoverable http errors
+            } else {
+                Log.debug(label: Self.LOG_TAG, "[\(Self.CLASS_NAME)<\(#function)>] - [Session(\(self.id))] \(eventType) request failed with unrecoverable HTTP error:(\(connection.responseMessage ?? "")) response code:(\(connection.responseCode ?? -1)). Request will be dropped.")
+                return false
             }
         }
+
+        return false
```
Comment (kevinlind) [inline AEPMedia/Sources/MediaCollection/MediaRealTimeSession.swift:234]: Add a trace log here. I'm assuming we shouldn't arrive here, but log out here that the sessionStart response has a connection without an error or responseCode, so we are dropping the request.

**Review by kevinlind — APPROVED** [review]
Looks good. Added a few minor comments to add more logging.

File: AEPMedia/Sources/MediaCollection/MediaRealTimeSession.swift (hunk 1 above)
Comment (addb) [reply AEPMedia/Sources/MediaCollection/MediaRealTimeSession.swift:208]: Connection is never nil. We invoke error from sendHit when we failed to get or the parse the sessionId from the server even when network request was successful with 200. So this error also should not occur in ideal situation.

**Review by kevinlind — APPROVED** [review]

**Pull request merged**